text/0165-instrumentation-api.md

```diff
@@ -0,0 +1,236 @@
+
+- `void end(Context context, REQUEST request, RESPONSE response, Throwable error)`: The method to
+invoke at the end of an operation. The `Context` returned from `start` is passed in, as well as the
+library-specific request and response objects and any error if present. The implementation retrieves
+a span from the context, invokes all extractors and request listeners, records the error and status,
+and ends the span.
+
+Many libraries offer a means of intercepting operations, for example a servlet filter, gRPC
+interceptor, etc. In these cases, a library author will simply create an implementation of this
+interceptor with an `Instrumenter` and call the methods. For example, in a servlet filter, the code
+would simply be something like
+
+```java
+class TracingServletFilter implements Filter {
+  private final Instrumenter instrumenter;
+
+  TracingServletFilter(Instrumenter instrumenter) {
+    this.instrumenter = instrumenter;
+  }
+
+  public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain) {
+    Context context = instrumenter.start(Context.current(), request);
+    Throwable error = null;
+    try (Scope ignored = context.makeCurrent()) {
+      chain.doFilter(request, response);
+    } catch (Throwable t) {
+      error = t;
+    }
+    instrumenter.end(context, request, response, error);
+  }
+}
+```
+
+### AttributesExtractor
+
+`AttribuesExtractor` allows populating attributes from a library's request and response domain
+objects.
+
+The `AttributesExtractor` interface itself is two callbacks that are called from `start` and `end`
+of the `Instrumenter`.
+
+- `onStart(AttributesBuilder attributes, REQUEST request)`: Populates attributes with information
+from the domain specific request object at the beginning of a request.
+
+- `onEnd(AttributesBuilder attributes, REQUEST request, RESPONSE response)`: Populates attributes
+with information from the domain specific request and response object at the end of a request.
+
+The main reason we must pass `request` in `onEnd` as well is because for many frameworks, information
+is contained in the request object but not available until the end, for example the `net.peer.ip`
+remote address for a client request.
+
+Additional interfaces are provided for `AttributesExtractor` corresponding to each semantic convention
+namespace.
+
+For example, `HttpAttributesExtractor` defines an interface with one method per HTTP semantic convention.
+The implementation of `onStart` and `onEnd` populate the attributes with the result of these methods,
+which operate on the request domain object. In Java it looks like [this](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/http/HttpAttributesExtractor.java).
+
+An implementation of each semantic convention method is required, i.e., no default implementation of the
```

Comment:
The Instrumentation API, at least in this case by using abstract methods in Java. Indeed this may not translate to other languages but even without the strict requirement, it could be more discoverable and ensure filling in attributes compared to `setAttribute` I think

Reply:
Sorry - let me rephrase. Are we saying here that we expect instrumentation authors to implement all of the abstract methods, even if they don't use them? Or that we expect the people who actually _build_ the instrumentation API _itself_ to implement no-op implementations of all attributes?

I do like the focus on explicitness, in general, but whether or not I personally like this part of the idea depends on who we expect to be implementing these methods, and when. 😄